src/integrity-checker.js

```diff
@@ -117,12 +199,38 @@ export default class InstallationIntegrityChecker {
       };
     }
 
-    const actual = this._generateIntegrityHash(normalizedLockSource, patterns, flags);
-    const expected = (await fs.readFile(loc.locationPath)).trim();
+    const actual = await this._generateIntegrityFile(
+      lockfile,
+      patterns,
+      Object.assign({}, {checkFiles: false}, flags), // don't generate files when checking, we check the files below
+      loc.locationFolder);
+    const expectedRaw = await fs.readFile(loc.locationPath);
+    let expected: IntegrityFile;
+    try {
+      expected = JSON.parse(expectedRaw);
+    } catch (e) {
+      // ignore JSON parsing for legacy text integrity files compatibility
+    }
+    let integrityMatches;
+    if (expected) {
+      integrityMatches = this._compareIntegrityFiles(actual, expected);
+      if (flags.checkFiles && expected.files.length > 0) {
+        // TODO we may want to optimise this check by checking only for package.json files on very large trees
```

Comment:
Yeah, I think we will evolve this solution once we understand how it is going to be used.
mtime makes sense especially that we already use it in fs.copy.
Maybe just a map {path: mtime} or {path: {size, mtime, or the whole stat}}.

I could not make a decision at the time of writing so decided to go there if we need it.